exchange/bitswap/peermanager.go

```diff
@@ -0,0 +1,224 @@
+package bitswap
+
+import (
+	"sync"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	blocks "github.com/ipfs/go-ipfs/blocks"
+	bsmsg "github.com/ipfs/go-ipfs/exchange/bitswap/message"
+	bsnet "github.com/ipfs/go-ipfs/exchange/bitswap/network"
+	peer "github.com/ipfs/go-ipfs/p2p/peer"
+	u "github.com/ipfs/go-ipfs/util"
+	// TODO move this to peerstore
+)
+
+type PeerManager struct {
+	receiver bsnet.Receiver
+
+	incoming   chan *msgPair
+	connect    chan peer.ID
+	disconnect chan peer.ID
+	cancels    chan cancellation
+
+	peers map[peer.ID]*msgQueue
+
+	network bsnet.BitSwapNetwork
+}
+
+func NewPeerManager(network bsnet.BitSwapNetwork) *PeerManager {
+	return &PeerManager{
+		incoming:   make(chan *msgPair, 10),
+		connect:    make(chan peer.ID, 10),
+		disconnect: make(chan peer.ID, 10),
+		cancels:    make(chan cancellation, 10),
+		peers:      make(map[peer.ID]*msgQueue),
+		network:    network,
+	}
+}
+
+type msgPair struct {
+	to  peer.ID
+	msg bsmsg.BitSwapMessage
+}
+
+type msgQueue struct {
+	p peer.ID
+
+	lk     sync.Mutex
+	wlmsg  bsmsg.BitSwapMessage
+	blocks map[u.Key]*blocks.Block
+
+	work chan struct{}
+	done chan struct{}
+}
+
+func (pm *PeerManager) runQueue(mq *msgQueue) {
+	for {
+		select {
+		case <-mq.work: // there is work to be done
+			err := pm.network.ConnectTo(context.TODO(), mq.p)
+			if err != nil {
+				log.Error(err)
+				// TODO: cant connect, what now?
+			}
+
+			// grab messages from queue
+			mq.lk.Lock()
+			wlm := mq.wlmsg
+			blks := mq.blocks
+			mq.wlmsg = nil
+			mq.blocks = make(map[u.Key]*blocks.Block)
```

Comment:
yeah, there isnt an easy way to clear a map, so i just created a new one. The idea is that i'm grabbing the blocks to be sent, and then sending them. I'm thinking now, that making it a slice is probably for the best, lookups on a small array isnt super costly, especially if we dont do it often (only upon a 'cancel block' from that peer), and its probably cheaper than all the map creation this incurs.

I'm also thinking about moving this into two segments, a path for wantlist updates, and a path for outgoing blocks, which makes it easier to cancel outgoing blocks at the last second.